### PR TITLE
feat: NewEvent provides a new item for setting pre-defined values.

### PR DESCRIPTION
**Pull request by hdamr**
## Description

Provide new item in `NewEvent`. Inside of event handler for `NewEvent`,  when we call `e.getItem()`, it used to return `null` value, which wasn't any useful. This PR modifies the event object of `NewEvent` to include an empty instance of type of the crud, so that end user can provide some default values to newly added objects via CRUD component.

Fixes https://github.com/vaadin/flow-components/issues/1105

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs-beta/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: vaadin-crud-flow-parent/vaadin-crud-flow-demo/src/main/java/com/vaadin/flow/component/crud/demo/CrudView.java [hunk 1]
```diff
@@ -76,7 +76,10 @@ private void basicCrud() {
         crud.setDataProvider(dataProvider);
         crud.addSaveListener(e -> dataProvider.persist(e.getItem()));
         crud.addDeleteListener(e -> dataProvider.delete(e.getItem()));
-
+        crud.addNewListener(e -> {
```
Comment (alvarezguille) [inline vaadin-crud-flow-parent/vaadin-crud-flow-demo/src/main/java/com/vaadin/flow/component/crud/demo/CrudView.java:79]: Is this change needed? This view is not used in tests

File: vaadin-crud-flow-parent/vaadin-crud-flow-integration-tests/src/test/java/com/vaadin/flow/component/crud/test/BasicUseIT.java [hunk 2]
```diff
@@ -131,7 +132,24 @@ public void crudAndGeneratedGridReactToThemeVariantChanges() {
...
+        crud.getEditorCancelButton().click();
+        Assert.assertEquals(
+                "Cancel: Person{id=0, firstName='firstName', lastName='lastName'}",
+                getLastEvent());
+    }
+
     private ButtonElement getTestButton(String id) {
         return $(ButtonElement.class).onPage().id(id);
     }
+
```
Comment (alvarezguille) [inline vaadin-crud-flow-parent/vaadin-crud-flow-integration-tests/src/test/java/com/vaadin/flow/component/crud/test/BasicUseIT.java:150]: This is overriding a method from AbstractParallelTest with the same implementation. It could be removed

File: vaadin-crud-flow-parent/vaadin-crud-flow-integration-tests/src/test/java/com/vaadin/flow/component/crud/test/BasicUseIT.java [hunk 3]
```diff
@@ -131,7 +132,24 @@ public void crudAndGeneratedGridReactToThemeVariantChanges() {
         Assert.assertNotEquals("no-border", grid.getAttribute("theme"));
     }
 
+    @Test
```
Comment (alvarezguille) [inline vaadin-crud-flow-parent/vaadin-crud-flow-integration-tests/src/test/java/com/vaadin/flow/component/crud/test/BasicUseIT.java:135]: This test could be in `EventHandlingIT` which is specific to events.

File: vaadin-crud-flow-parent/vaadin-crud-flow-integration-tests/src/test/java/com/vaadin/flow/component/crud/test/BasicUseIT.java [hunk 4]
```diff
@@ -131,7 +132,24 @@ public void crudAndGeneratedGridReactToThemeVariantChanges() {
...
+    public void newEventItem() {
+        CrudElement crud = $(CrudElement.class).waitForFirst();
+        $("#newEventListener").onPage().first().click();
```
Comment (alvarezguille) [inline vaadin-crud-flow-parent/vaadin-crud-flow-integration-tests/src/test/java/com/vaadin/flow/component/crud/test/BasicUseIT.java:138]: As a general comment using `$("#newEventListener")` with a selector might work but it's not a supported case.
In this case  the implementation in `getTestButton` is preferable, you can use that helper instead.

File: vaadin-crud-flow-parent/vaadin-crud-flow-integration-tests/src/test/java/com/vaadin/flow/component/crud/test/EventHandlingIT.java [hunk 5]
```diff
@@ -161,7 +163,7 @@ public void saveTest() {
                 $(GridElement.class).first().getCell(0, 2).getText());
 
         // Valid input
-        lastNameField.setValue("Oladeji");
+        lastNameField.setValue("Obbbbbbbbbbb");
```
Comment (alvarezguille) [inline vaadin-crud-flow-parent/vaadin-crud-flow-integration-tests/src/test/java/com/vaadin/flow/component/crud/test/EventHandlingIT.java:166]: Are these changes needed?

**Review by alvarezguille — CHANGES_REQUESTED** [review]
Added some general comments about TestBench usage and ITs. 
Didn't review the new feature itself, @DiegoCardoso please take a look.

File: vaadin-crud-flow-parent/vaadin-crud-flow-integration-tests/src/test/java/com/vaadin/flow/component/crud/test/BasicUseIT.java (hunk 2 above)
Comment (hdamr) [reply vaadin-crud-flow-parent/vaadin-crud-flow-integration-tests/src/test/java/com/vaadin/flow/component/crud/test/BasicUseIT.java:138]: There was no such method in `AbstractParallelTest`.

File: vaadin-crud-flow-parent/vaadin-crud-flow/src/main/java/com/vaadin/flow/component/crud/Crud.java [hunk 6]
```diff
@@ -893,6 +895,11 @@ private NewEvent(Crud<E> source, boolean fromClient, E item,
             this.item = item;
         }
 
+        /**
+         * Gets new item being created
+         * 
+         * @return a new object of bean type
```
Comment (sissbruecker) [inline vaadin-crud-flow-parent/vaadin-crud-flow/src/main/java/com/vaadin/flow/component/crud/Crud.java:901]: Nit: instance is a better term in Java-land

File: vaadin-crud-flow-parent/vaadin-crud-flow/src/test/java/com/vaadin/flow/component/crud/CrudTest.java [hunk 7]
```diff
@@ -51,6 +49,22 @@ public void itemAvailableInAllEvents() {
                 .forEach(e -> ComponentUtil.fireEvent(systemUnderTest, e));
     }
 
+    @Test
+    public void sameItemInNewEvent() {
```
Comment (sissbruecker) [inline vaadin-crud-flow-parent/vaadin-crud-flow/src/test/java/com/vaadin/flow/component/crud/CrudTest.java:53]: I don't quite understand the purpose of this test.

File: vaadin-crud-flow-parent/vaadin-crud-flow/src/test/java/com/vaadin/flow/component/crud/CrudTest.java [hunk 8]
```diff
@@ -29,9 +29,7 @@ public void itemAvailableInAllEvents() {
                 .addDeleteListener(e -> Assert.assertNotNull(e.getItem()));
         systemUnderTest.addEditListener(e -> Assert.assertNotNull(e.getItem()));
         systemUnderTest.addSaveListener(e -> Assert.assertNotNull(e.getItem()));
-
-        // Client side new should not come with an item
-        systemUnderTest.addNewListener(e -> Assert.assertNull(e.getItem()));
+        systemUnderTest.addNewListener(e -> Assert.assertNotNull(e.getItem()));
```
Comment (sissbruecker) [inline vaadin-crud-flow-parent/vaadin-crud-flow/src/test/java/com/vaadin/flow/component/crud/CrudTest.java:32]: I think we also want to make sure that this item in the new event is the same that is in the editor. If it would be a separate instance for some reason, then initializing it would not affect the editor.

File: vaadin-crud-flow-parent/vaadin-crud-flow-integration-tests/src/test/java/com/vaadin/flow/component/crud/test/BasicUseIT.java (hunk 2 above)
Comment (sissbruecker) [reply vaadin-crud-flow-parent/vaadin-crud-flow-integration-tests/src/test/java/com/vaadin/flow/component/crud/test/BasicUseIT.java:138]: This comment is still valid, the method that is being overridden is here: https://github.com/vaadin/flow-components/blob/master/vaadin-crud-flow-parent/vaadin-crud-flow-integration-tests/src/test/java/com/vaadin/flow/component/crud/test/AbstractParallelTest.java#L7-L9

File: vaadin-crud-flow-parent/vaadin-crud-flow-integration-tests/src/test/java/com/vaadin/flow/component/crud/test/EventHandlingIT.java [hunk 9]
```diff
@@ -217,6 +219,24 @@ public void invalidFieldsIndicatedOnSave() {
         Assert.assertTrue(lastNameField.hasAttribute("invalid"));
     }
 
+    @Test
+    public void newEventItem() {
```
Comment (sissbruecker) [inline vaadin-crud-flow-parent/vaadin-crud-flow-integration-tests/src/test/java/com/vaadin/flow/component/crud/test/EventHandlingIT.java:223]: Naming: It's not clear that this tests that the new item can be initialized / prepared in the new event listener.

File: vaadin-crud-flow-parent/vaadin-crud-flow/src/main/java/com/vaadin/flow/component/crud/Crud.java [hunk 10]
```diff
@@ -185,8 +185,10 @@ private void registerHandlers() {
                                 "Unable to instantiate new bean", ex);
                     }
 
-                    newListeners
-                            .forEach(listener -> listener.onComponentEvent(e));
+                    NewEvent modifiedEvent = new NewEvent(e.getSource(),
```
Comment (sissbruecker) [inline vaadin-crud-flow-parent/vaadin-crud-flow/src/main/java/com/vaadin/flow/component/crud/Crud.java:188]: Naming: `eventWithNewItem`?

**Review by sissbruecker — COMMENTED** [review]
Added some small comments.

However in general I think the whole approach of having to add a new event listener to initialize the new item is weird and difficult to discover. I think a better alternative would be if one could set an item factory in the Crud that takes care of generating / initializing new items. That would also allow to delegate the construction of instances to the user, which for example would allow to use classes that have constructor arguments.

File: vaadin-crud-flow-parent/vaadin-crud-flow/src/test/java/com/vaadin/flow/component/crud/CrudTest.java (hunk 7 above)
Comment (hdamr) [reply vaadin-crud-flow-parent/vaadin-crud-flow/src/test/java/com/vaadin/flow/component/crud/CrudTest.java:53]: The thing we wanted to test is that whatever value we set for new instance (in this case a particular field), shouldn't be changed or modified inside crud component, and we expect to receive the exact same value from `getItem().field`.

File: vaadin-crud-flow-parent/vaadin-crud-flow-integration-tests/src/test/java/com/vaadin/flow/component/crud/test/BasicUseIT.java [hunk 11]
```diff
@@ -9,6 +9,7 @@
 import com.vaadin.flow.component.button.testbench.ButtonElement;
 import com.vaadin.flow.component.crud.testbench.CrudElement;
 import com.vaadin.flow.component.grid.testbench.GridElement;
+import com.vaadin.flow.component.orderedlayout.testbench.VerticalLayoutElement;
```
Comment (sissbruecker) [inline vaadin-crud-flow-parent/vaadin-crud-flow-integration-tests/src/test/java/com/vaadin/flow/component/crud/test/BasicUseIT.java:12]: Unused import

**Review by sissbruecker — APPROVED** [review]

**Review by alvarezguille — APPROVED** [review]

**Review by DiegoCardoso — APPROVED** [review]
